### PR TITLE
Add Email Log

### LINKED ISSUE #3978: Log for outbound emails
The process of sending emails is quite secretive, with the background job silently chewing up errors until the mail is either sent out or fails after too many retries. This has two downsides: First of all it is pretty much unknown to the admin if an email has already been sent and second it prevents errors to be detected early on, because the error is silently suppressed via the queue system.

* [ ] Log all outbound emails.
* [ ] Provide some means to identify the type of emails.
  * [ ] Do not use the raw subject to prevent the leak of any sensitive data.
  * [ ] Possible solution: Some kind of generic identifier and optional id, e.g. `thread 123`, `conversation 456` or `contactForm`?
* [ ] Readable status with a highlighting or other kind of indicator for failed messages.
* [ ] Basic filter?

See https://community.woltlab.com/thread/287519-mail-versand-logging/

### PR DESCRIPTION
**Pull request by TimWolla**
ToDo:

- [ ] Set additional Message-IDs.
- [x] Filter within the log page.

Bildschirmfoto von 2021-02-15 14-30-23
Bildschirmfoto von 2021-02-15 14-30-39
Bildschirmfoto von 2021-02-15 14-30-51

-----------

Resolves #3978

### CODE REVIEW

File: wcfsetup/install/files/lib/system/background/job/EmailDeliveryBackgroundJob.class.php [hunk 1]
```diff
@@ -59,6 +74,54 @@ public function __construct(Email $email, Mailbox $envelopeFrom, Mailbox $envelo
...
+    /**
+     * Creates the email log entry.
+     */
+    private function createLog(): EmailLogEntry
+    {
+        return (new EmailLogEntryAction([], 'create', [
+            'data' => [
+                'time' => \TIME_NOW,
+                'messageID' => $this->email->getMessageID(),
+                'subject' => StringUtil::truncate($this->email->getSubject(), 30),
```
Comment (dtdesign) [inline wcfsetup/install/files/lib/system/background/job/EmailDeliveryBackgroundJob.class.php:90]:
```suggestion
                'subject' => StringUtil::truncate($this->email->getSubject(), 20),
```
Rationale: Trim to 20 characters to reduce the likelihood of exposing sensitive or confidential data in the mail subject.

File: wcfsetup/install/files/lib/system/email/transport/SmtpEmailTransport.class.php [hunk 2]
```diff
@@ -468,6 +468,8 @@ public function deliver(Email $email, Mailbox $envelopeFrom, Mailbox $envelopeTo
             return $item;
         }, \explode("\r\n", $email->getEmail()))) . "\r\n");
         $this->write(".");
-        $this->read([250]);
+        [$code, $message] = $this->read([250]);
```
Comment (dtdesign) [inline wcfsetup/install/files/lib/system/email/transport/SmtpEmailTransport.class.php:471]:
```suggestion
        [, $message] = $this->read([250]);
```
Rationale: `$code` was never used and this makes it clear that this is intentional.

**Review by dtdesign — CHANGES_REQUESTED** [review]

**Comment by TimWolla** [comment]
@dtdesign I've added the filter within the newest commit. PTAL. If it is good then this can be merged, the missing Message-IDs can be added later on demand.

**Review by dtdesign — APPROVED** [review]

**Comment by TimWolla** [comment]
Rebased to squash and reorder some commits, waiting for GitHub Actions to be green, then I'll merge.

**Pull request merged**